drag_and_drop_v2/drag_and_drop_v2.py

```diff
@@ -416,6 +377,50 @@ def _get_ngettext(self):
         else:
             return ngettext_fallback
 
+    def _validate_do_attempt(self):
+        if self.mode != self.ASSESSMENT_MODE:
+            raise JsonHandlerError(400, _("do_attempt handler should only be called for assessment mode"))
+        if not self.attemps_remain:
+            raise JsonHandlerError(409, _("Max number of attempts reached"))
+
+    def _get_do_attempt_feedback(self):
+        ngettext = self._get_ngettext()
+
+        required, placed, correct = self._get_item_raw_stats()
+        placed_ids, correct_ids = set(placed), set(correct)
+        missing_ids = set(required) - set(placed)
+        misplaced_ids = placed_ids - correct_ids
+
+        correct_count, misplaced_count, missing_count = len(correct_ids), len(misplaced_ids), len(missing_ids)
+
+        feedback_msgs = [
+            ngettext(
+                FeedbackMessages.CORRECTLY_PLACED_SINGULAR_TPL,
+                FeedbackMessages.CORRECTLY_PLACED_PLURAL_TPL,
+                correct_count
+            ).format(correct_count=correct_count),
+            ngettext(
+                FeedbackMessages.MISPLACED_SINGULAR_TPL,
+                FeedbackMessages.MISPLACED_PLURAL_TPL,
+                misplaced_count
+            ).format(misplaced_count=misplaced_count),
+            ngettext(
+                FeedbackMessages.NOT_PLACED_REQUIRED_SINGULAR_TPL,
+                FeedbackMessages.NOT_PLACED_REQUIRED_PLURAL_TPL,
+                missing_count
+            ).format(missing_count=missing_count)
```

Comment:
@e-kolpakov We might want to reword this message, or maybe just remove it when `missing_count == 0`, because "Not placed 0 required items" is more confusing than it is helpful.